src/main/java/org/mariadb/jdbc/internal/failover/impl/MastersSlavesListener.java

```diff
@@ -434,80 +434,88 @@ public void lockAndSwitchSecondary(Protocol newSecondaryProtocol) throws QueryEx
      */
     @Override
     public void switchReadOnlyConnection(Boolean mustBeReadOnly) throws QueryException {
-        if (mustBeReadOnly != currentReadOnlyAsked.get() && currentProtocol.inTransaction()) {
+        if (mustBeReadOnly != currentReadOnlyAsked && currentProtocol.inTransaction()) {
             throw new QueryException("Trying to set to read-only mode during a transaction");
         }
         checkWaitingConnection();
-        if (currentReadOnlyAsked.compareAndSet(!mustBeReadOnly, mustBeReadOnly)) {
```

Comment:
@jentfoo I'm not sure why you switched from an AtomicBoolean to a volatile and a synchronized object combination. If you are sure that AtomicBoolean is defunct, I would say remove it. But I don't think this is the case. I would recommend leaving the AtomicBoolean as is, until the problem is further investigated.

Further reads on AtomicBoolean versus volatile variables in java: 
http://stackoverflow.com/a/3787435/367988
http://stackoverflow.com/a/3787093/367988